### PR TITLE
add source comments to input_drivers.h

### PR DESCRIPTION
**Pull request by markwkidd**
This my latest work to add documentation for the input drivers.

I have started using a minimal Doxygen style for the function prototype comments since a lot of the existing comments were close to that style already.

I'll follow up with a few questions I have about documenting some of these function.

### CODE REVIEW

File: input/input_driver.h [hunk 1]
```diff
@@ -89,122 +91,204 @@ enum input_action
...
+    */
    enum msg_hash_enums enum_idx;
+
+   /**
+    * TODO
+    */
    enum retro_key key;
 
+   /**
+    * TODO
```
Comment (RobLoach) [inline input/input_driver.h:135]: All of these TODO notes don't really add any value.

**Comment by markwkidd** [comment]
Yes I have a few questions to clear up, but I didn't finish writing them up
yesterday. I should mark this as a draft!

**Comment by markwkidd** [comment]
Changed this PR to a draft. I want to ask a few questions, and then hopefully fill in all the remaining todos.

**Comment by RobLoach** [comment]
While a bunch of these changes are valid, adding a bunch of stub documentation adds bloat without any value. You can change the doxygen settings to adjust more to what Retroarch uses. Like the line break number.

**Comment by markwkidd** [comment]
For what it's worth, this should all be 80 characters wide already.

I do intend to fill in every stub 😅

**Comment by markwkidd** [comment]
I've been able to clarify some of my questions, leading me to identify two sections of this header which I believe are better located in other places.

I have created PRs: https://github.com/libretro/RetroArch/pull/12617 and https://github.com/libretro/RetroArch/pull/12618. 

Meanwhile I hope it makes sense for me to leave this PR as a draft another day or two while I keep filling in the remaining blank spots.

**Comment by markwkidd** [comment]
Maybe it was predictable, but I am going to narrow the scale of this PR. I have removed the remaining TODO comments -- everything in the PR is actual documentation. I would be grateful for a review and hopefully a merge.

I will continue to work on adding more documentation. Just under a new PR! 🏓

**Comment by markwkidd** [comment]
@RobLoach do you have any time to look at this PR again? it's cleaned up now with no TODOs. The latter part of the file is split into a separate PR https://github.com/libretro/RetroArch/pull/12661 which also hopefully makes it easier to review

File: input/input_driver.h [hunk 2]
```diff
@@ -45,35 +45,55 @@ RETRO_BEGIN_DECLS
...
 
-   /* Default key binding value -
-    * for resetting bind to default */
+   /**
+    * Default key binding value (for resetting bind).
+    */
    uint16_t def_joykey;
 
+   /**
+    * Is this bind valid?
```
Comment (RobLoach) [inline input/input_driver.h:95]:
``` suggestion
    * Determines whether or not the binding is valid.
```

File: input/input_driver.h [hunk 3]
```diff
@@ -169,83 +186,73 @@ struct input_driver
          unsigned port, unsigned device, unsigned index, unsigned id);
 
    /**
-    * free: Frees the input struct.
+    * Frees the input struct.
     * 
-    * PARAMETERS
-    * @data:   The input state struct
+    * @param data  The input state struct
```
Comment (RobLoach) [inline input/input_driver.h:191]:
``` suggestion
    * @param data The input state struct.
```

File: input/input_driver.h [hunk 4]
```diff
@@ -169,83 +186,73 @@ struct input_driver
...
-    * @effect: Sensor action
-    * @rate:   Sensor rate update
+    * @param data    The input state struct
+    * @param port
+    * @param effect  Sensor action
+    * @param rate    Sensor rate update
     * 
-    * RETURNS
-    * Boolean true if the operation is successful. 
+    * @return       true if the operation is successful.
```
Comment (RobLoach) [inline input/input_driver.h:205]:
``` suggestion
    * @return true if the operation is successful.
```

File: input/input_driver.h [hunk 5]
```diff
@@ -417,46 +399,64 @@ const input_device_driver_t *input_joypad_init_driver(
...
+
+
+/**
+ * line_complete callback (when carriage return is pressed)
+ *
+ * @param userdata  ??
+ * @param line      the line of input, which can be NULL.
  **/
-typedef void (*input_keyboard_line_complete_t)(void *userdata,
+typedef void (*input_keyboard_line_complete_t)(void *userdata,
```
Comment (RobLoach) [inline input/input_driver.h:451]:
``` suggestion
typedef void (*input_keyboard_line_complete_t)(void *userdata,
```

File: input/input_driver.h [hunk 6]
```diff
@@ -417,46 +399,64 @@ const input_device_driver_t *input_joypad_init_driver(
...
- * Calls back after return is pressed with the completed line.
- * Line can be NULL.
+#endif /* HAVE_HID */
+/*****************************************************************************/
+
+
+/**
+ * line_complete callback (when carriage return is pressed)
+ *
+ * @param userdata  ??
```
Comment (RobLoach) [inline input/input_driver.h:448]:
``` suggestion
 * @param userdata User data which will be passed to subsequent callbacks.
```

File: input/input_driver.h [hunk 7]
```diff
@@ -417,46 +399,64 @@ const input_device_driver_t *input_joypad_init_driver(
...
+ * @param line      the line of input, which can be NULL.
  **/
-typedef void (*input_keyboard_line_complete_t)(void *userdata,
+typedef void (*input_keyboard_line_complete_t)(void *userdata, 
       const char *line);
 
+/**
+ * Callback for keypress events
+ * 
+ * @param userdata  ??
```
Comment (RobLoach) [inline input/input_driver.h:457]:
``` suggestion
 * @param userdata The user data that was passed through from the keyboard press callback.
```

File: input/input_driver.h [hunk 8]
```diff
@@ -503,38 +587,142 @@ unsigned *input_config_get_device_ptr(unsigned port);
...
 unsigned input_config_get_device_name_index(unsigned port);
 
-/* TODO/FIXME: This is required by linuxraw_joypad.c
- * and parport_joypad.c. These input drivers should
- * be refactored such that this dubious low-level
- * access is not required */
+
+
+/**
+ * Deprecated: input_config_get_device_name_ptr is required by linuxraw_joypad
```
Comment (RobLoach) [inline input/input_driver.h:636]:
``` suggestion
 * Retrieve the device name char pointer.
 *
 * @deprecated input_config_get_device_name_ptr is required by linuxraw_joypad
```

File: input/input_driver.h [hunk 9]
```diff
@@ -503,38 +587,142 @@ unsigned *input_config_get_device_ptr(unsigned port);
...
+ * @param port
+ *
+ * @return a pointer to the device name on the specified port
+ */
 char *input_config_get_device_name_ptr(unsigned port);
-size_t input_config_get_device_name_size(unsigned port);
 
-const struct retro_keybind *input_config_get_bind_auto(unsigned port, unsigned id);
+/**
+ * Deprecated: input_config_get_device_name_size is required by linuxraw_joypad
```
Comment (RobLoach) [inline input/input_driver.h:647]:
``` suggestion
 * Get the size of the device name.
 *
 * @deprecated input_config_get_device_name_size is required by linuxraw_joypad
```

File: input/input_driver.h [hunk 10]
```diff
@@ -267,134 +274,109 @@ struct rarch_joypad_driver
...
 const char* config_get_joypad_driver_options(void);
 
 /**
- * input_joypad_init_driver:
- * @ident                           : identifier of driver to initialize.
- *
- * Initialize a joypad driver of name @ident.
+ * Initialize a joypad driver of name ident. If ident points to NULL or a 
+ * zero-length string, equivalent to calling input_joypad_init_first().
+ *
```
Comment (RobLoach) [inline input/input_driver.h:363]:
``` suggestion
 *
```

File: input/input_driver.h [hunk 11]
```diff
@@ -417,46 +399,64 @@ const input_device_driver_t *input_joypad_init_driver(
...
  **/
 const char* config_get_hid_driver_options(void);
 
 /**
- * input_hid_init_first:
- *
  * Finds first suitable HID driver and initializes.
- *
- * Returns: HID driver if found, otherwise NULL.
+ *
```
Comment (RobLoach) [inline input/input_driver.h:423]:
``` suggestion
 *
```

File: input/input_driver.h [hunk 12]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
       uint16_t mod, unsigned device);
 
-
-/* Set input_device_info */
+/**
+ * Set the name of the device in the specified port
+ * 
+ * @param port
+ *
+ * @return the device name
```
Comment (RobLoach) [inline input/input_driver.h:485]:
``` suggestion
```

File: input/input_driver.h [hunk 13]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_name(unsigned port, const char *name);
+
+/**
+ * Set the formatted "display name" of the device in the specified port
+ * 
+ * @param port
+ *
+ * @return display name for the device
```
Comment (RobLoach) [inline input/input_driver.h:494]:
``` suggestion
```

File: input/input_driver.h [hunk 14]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_display_name(unsigned port, const char *name);
+
+/**
+ * Set the configuration path for the device in the specified port
+ * 
+ * @param port
+ *
+ * @return configuration path for the device
```
Comment (RobLoach) [inline input/input_driver.h:503]:
``` suggestion
 * @param path The path of the device config.
```

File: input/input_driver.h [hunk 15]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_config_path(unsigned port, const char *path);
+
+/**
+ * Set the configuration name for the device in the specified port
+ * 
+ * @param port
+ *
+ * @return configuration name for the device
```
Comment (RobLoach) [inline input/input_driver.h:512]:
``` suggestion
 * @param name The name of config to set.
```

File: input/input_driver.h [hunk 16]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_config_name(unsigned port, const char *name);
+
+/**
+ * Set the joypad driver for the device in the specified port
+ * 
+ * @param port
+ *
+ * @return joypad driver
```
Comment (RobLoach) [inline input/input_driver.h:521]:
``` suggestion
 * @param driver The driver of which to set the given port to.
```

File: input/input_driver.h [hunk 17]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_joypad_driver(unsigned port, const char *driver);
+
+/**
+ * Set the vendor ID (vid) for the device in the specified port
+ * 
+ * @param port
+ *
+ * @return vendor ID (vid)
```
Comment (RobLoach) [inline input/input_driver.h:530]:
``` suggestion
 * @param vid The vid of which to set the given device port to.
```

File: input/input_driver.h [hunk 18]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_vid(unsigned port, uint16_t vid);
+
+/**
+ * Set the pad ID (pid) for the device in the specified port
+ * 
+ * @param port
+ *
+ * @return pad ID (pid)
```
Comment (RobLoach) [inline input/input_driver.h:539]:
``` suggestion
 * @param pid The pid of which to set the given device port to.
```

File: input/input_driver.h [hunk 19]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_pid(unsigned port, uint16_t pid);
+
+/**
+ * Sets the autoconfigured flag for the device in the specified port
+ *
+ * @param port
+ *
+ * @return true if the device is autoconfigured
```
Comment (RobLoach) [inline input/input_driver.h:548]:
``` suggestion
 * @param autoconfigured Whether or nor the device is configured automatically.
```

File: input/input_driver.h [hunk 20]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_autoconfigured(unsigned port, bool autoconfigured);
+
+/**
+ * Sets the name index number for the device in the specified port
+ *
+ * @param port
+ *
+ * @return the name index for this device
```
Comment (RobLoach) [inline input/input_driver.h:557]:
``` suggestion
 * @param name_index The name index of which to set the device to.
```

File: input/input_driver.h [hunk 21]
```diff
@@ -466,30 +466,114 @@ struct input_keyboard_ctx_wait
...
+ */
 void input_config_set_device_name_index(unsigned port, unsigned name_index);
 
+/**
+ * Sets the device type of the specified port
+ * 
+ * @param port
+ * @param id    device type (eg RETRO_DEVICE_JOYPAD, RETRO_DEVICE_LIGHTGUN)
```
Comment (RobLoach) [inline input/input_driver.h:565]:
``` suggestion
 * @param id The device type (RETRO_DEVICE_JOYPAD, RETRO_DEVICE_MOUSE, etc)
```

**Review by RobLoach — CHANGES_REQUESTED** [review]
Looks great! Gives a good understanding of how the API works, and leverages doxygen standards pretty well. Nice work! A few suggestions :wink:

**Comment by markwkidd** [comment]
Thank you @RobLoach , it's obvious you looked that this closely. I've accepted your suggestions, and force-pushed the PR to squash the commits.

**Pull request merged**